### PR TITLE
Implement OData v4 extension for data aggregation

### LINKED ISSUE #490: Introduce $apply for fluent interface
I've recently been playing with $apply and have run into the limitation of Simple.OData.Client not have a Fluent interface to utilize it (or the documentation doesn't exist). I've found a way around it, but it has some caveats that don't quite match the OData spec, but it works. Just hoping that adding $apply, along with the aggregation features could be a future enhancement. (Actually looking to contribute that...just not sure where to begin)

### PR DESCRIPTION
**Pull request by maxbruecken**
This (partially) implements the OData Extension for Data Aggregation as described [here](http://docs.oasis-open.org/odata/odata-data-aggregation-ext/v4.0/odata-data-aggregation-ext-v4.0.html). Due to complexity of the extension only the transformations `filter`, `aggregate` and `groupby` are supported yet.

The extension can be used either strictly typed e.g.
```
var productsGroupedByCategory = _client
                .WithExtensions()
                .For<Product>()
                .Apply(b => b.GroupBy((x, a) => new
                {
                    x.Category.CategoryName,
                    AverageUnitPrice = a.Average(x.UnitPrice)
                }))
                .OrderBy(x => x.AverageUnitPrice)
                .FindEntriesAsync();
```
or dynamically e.g.
```
var x = ODataDynamic.Expression;
var b = ODataDynamicDataAggregation.Builder;
var a = ODataDynamicDataAggregation.AggregationFunction;
var productsGroupedByCategory = _client
                .WithExtensions()
                .For(x.Product)
                .Apply(b.GroupBy(new
                {
                    x.Category.CategoryName,
                    AverageUnitPrice = a.Average(x.UnitPrice)
                }))
                .OrderBy(x.AverageUnitPrice)
                .FindEntriesAsync();
```
This pull request should fix #490 and probably #748, #415

### CODE REVIEW

File: src/Simple.OData.Client.Core/Extensions/DictionaryExtensions.cs [hunk 1]
```diff
@@ -281,5 +287,21 @@ private static ODataEntry CreateODataEntry(IDictionary<string, object> source, I
             property.SetValue(instance, dynamicProperties, null);
             return dynamicProperties;
         }
+
+        private static object CreateInstanceOfAnonymousType(IDictionary<string, object> source, Type type, ITypeCache typeCache)
+        {
+            var constructorParameterTypes = source.Values.Select(v => v.GetType()).ToArray();
```
Comment (watfordgnf) [inline src/Simple.OData.Client.Core/Extensions/DictionaryExtensions.cs:293]: I like this method, but it will be expensive on the hot path given three Select's and three ToArray's. All three collections could be exactly sized arrays as well.

It also relies on the fact that `Dictionary<K,V>` iterates in the order of insertion which is an implementation detail.

File: src/Simple.OData.Client.Core/Fluent/FluentClientBase.cs [hunk 2]
```diff
@@ -14,7 +13,7 @@ namespace Simple.OData.Client
     /// </summary>
     /// <typeparam name="T">The entity type.</typeparam>
     /// <typeparam name="FT"></typeparam>
-    public abstract partial class FluentClientBase<T, FT> : IFluentClient<T, FT>
+    public abstract class FluentClientBase<T, FT> : IFluentClient<T, FT>
```
Comment (watfordgnf) [inline src/Simple.OData.Client.Core/Fluent/FluentClientBase.cs:16]: Why was `partial` dropped?

File: src/Simple.OData.Client.Core/Fluent/FluentCommand.cs [hunk 3]
```diff
@@ -14,7 +14,7 @@ namespace Simple.OData.Client
     // For the same reason FluentClient is also declared as public
     // More: http://bloggingabout.net/blogs/vagif/archive/2013/08/05/we-need-better-interoperability-between-dynamic-and-statically-compiled-c.aspx
 
-    public partial class FluentCommand
+    public class FluentCommand
```
Comment (watfordgnf) [inline src/Simple.OData.Client.Core/Fluent/FluentCommand.cs:17]: Why was `partial` dropped?

File: src/Simple.OData.Client.V4.Adapter/Extensions/ExtendedBoundClient.cs [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+        }
+
+        public IExtendedBoundClient<TR> Apply<TR>(Func<IDataAggregation<T>, IDataAggregation<TR>> dataAggregation) where TR : class
+        {
+            var dataAggregationBuilder = new DataAggregationBuilder<T>(Session);
+            dataAggregation(dataAggregationBuilder);
+            AppendDataAggregationBuilder(dataAggregationBuilder);
+            if (typeof(T) != typeof(TR))
+            {
+                Session.Settings.IgnoreUnmappedProperties = true;
```
Comment (watfordgnf) [inline src/Simple.OData.Client.V4.Adapter/Extensions/ExtendedBoundClient.cs:28]: This seems out of place as now the entire session is ignoring unmapped properties.

File: src/Simple.OData.Client.V4.Adapter/Extensions/ExtendedBoundClient.cs [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+            AppendDataAggregationCommand(dataAggregationCommand);
+            Session.Settings.IgnoreUnmappedProperties = true;
+            return new ExtendedBoundClient<TR>(Session, Command);
+        }
+
+        public IExtendedBoundClient<T> Apply(DynamicDataAggregation dataAggregation)
+        {
+            Command.Details.Extensions[ODataLiteral.Apply] = dataAggregation.CreateBuilder();
+            Session.Settings.IgnoreUnmappedProperties = true;
+            Session.Settings.ReadUntypedAsString = false;
```
Comment (watfordgnf) [inline src/Simple.OData.Client.V4.Adapter/Extensions/ExtendedBoundClient.cs:50]: This seems out of place as now the entire session is reading untyped values as string.

Would it instead be better to instruct users who need this to set that up rather than changing the entire session settings?

File: src/Simple.OData.Client.V4.Adapter/Extensions/ExtendedODataClient.cs [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+        {
+            _baseClient = baseClient;
+        }
+
+        public IExtendedBoundClient<IDictionary<string, object>> For(string collectionName)
+        {
+            var client = new ExtendedBoundClient<IDictionary<string, object>>(_baseClient, _baseClient.Session);
+            client.For(collectionName);
+            _baseClient.Session.Settings.IgnoreUnmappedProperties = true;
+            _baseClient.Session.Settings.ReadUntypedAsString = false;
```
Comment (watfordgnf) [inline src/Simple.OData.Client.V4.Adapter/Extensions/ExtendedODataClient.cs:22]: I presume this works around some problem, but I'm not quite sure what, and it seems out of place.

File: src/Simple.OData.Client.V4.Adapter/ResponseReader.cs [hunk 7]
```diff
@@ -216,7 +216,7 @@ private ODataEntryAnnotations CreateAnnotations(ODataResource odataEntry)
             {
                 try
                 {
-                    id = odataEntry.Id.AbsoluteUri;
+                    id = odataEntry.Id?.AbsoluteUri;
```
Comment (watfordgnf) [inline src/Simple.OData.Client.V4.Adapter/ResponseReader.cs:219]: Why was this needed as part of this change?

**Review by watfordgnf — COMMENTED** [review]
Some questions and concerns, we're not using this extension yet so I'm not quite up to date on it.

File: src/Simple.OData.Client.V4.Adapter/ResponseReader.cs (hunk 7 above)
Comment (maxbruecken) [reply src/Simple.OData.Client.V4.Adapter/ResponseReader.cs:219]: OData service generates for aggregated entites ids that are nulls (s. http://docs.oasis-open.org/odata/odata-json-format/v4.0/errata03/os/odata-json-format-v4.0-errata03-os-complete.html#_Toc453766634) so this change was necessary to avoid NullReferenceException.

Comment (watfordgnf) [reply src/Simple.OData.Client.V4.Adapter/ResponseReader.cs:219]: Gotcha, so basically this line from the spec?
> If the entity is transient (i.e. cannot be read or updated), the odata.id annotation MUST appear and have the null value.

Comment (maxbruecken) [reply src/Simple.OData.Client.V4.Adapter/ResponseReader.cs:219]: yes, right

File: src/Simple.OData.Client.Core/Fluent/FluentClientBase.cs (hunk 2 above)
Comment (maxbruecken) [reply src/Simple.OData.Client.Core/Fluent/FluentClientBase.cs:16]: The implementation of FluentClientBase contains only this single part so `partial` is not necessary here. This change is not related to my other changes so I can revert this.

File: src/Simple.OData.Client.Core/Fluent/FluentCommand.cs (hunk 3 above)
Comment (maxbruecken) [reply src/Simple.OData.Client.Core/Fluent/FluentCommand.cs:17]: The implementation of FluentCommand contains only this single part so `partial` is not necessary here. This change is not related to my other changes so I can revert this.

File: src/Simple.OData.Client.V4.Adapter/ResponseReader.cs (hunk 7 above)
Comment (watfordgnf) [reply src/Simple.OData.Client.V4.Adapter/ResponseReader.cs:219]: If you wouldn't mind adding a comment for that spec link. As this project expands to cover more and more of the OData features it becomes harder to figure out exactly why certain design decisions are made.

File: src/Simple.OData.Client.Core/Extensions/DictionaryExtensions.cs (hunk 1 above)
Comment (maxbruecken) [reply src/Simple.OData.Client.Core/Extensions/DictionaryExtensions.cs]: I have refactored the creation of instances of anonymous type to avoid using `.Select(...).ToArray()` multiple times

File: src/Simple.OData.Client.V4.Adapter/Extensions/ExtendedBoundClient.cs (hunk 4 above)
Comment (maxbruecken) [reply src/Simple.OData.Client.V4.Adapter/Extensions/ExtendedBoundClient.cs]: I have compeletely removed changing of session settings and added a remark to the xml docs to instruct users to set this setting if necessary

File: src/Simple.OData.Client.V4.Adapter/Extensions/ExtendedODataClient.cs (hunk 6 above)
Comment (maxbruecken) [reply src/Simple.OData.Client.V4.Adapter/Extensions/ExtendedODataClient.cs:22]: I have compeletely removed changing of session settings and added a remark to the xml docs to instruct users to set this setting if necessary

File: src/Simple.OData.Client.V4.Adapter/Extensions/ExtendedBoundClient.cs (hunk 5 above)
Comment (maxbruecken) [reply src/Simple.OData.Client.V4.Adapter/Extensions/ExtendedBoundClient.cs]: I have compeletely removed changing of session settings and added a remark to the xml docs to instruct users to set this setting if necessary

**Comment by object** [comment]
Thank you for this PR. These extensions look very well thought and properly implemented. Special thanks for unit tests.
